File: backends/fastapi/app/services/favorite_service.py
"""收藏 / 取消收藏：维护 favorites 表与 capsules.favorite_count 的一致性。

关键约束：
  - favorite_count ± 1 与 favorites 行变更必须同事务
  - 并发安全：favorite_count 用 SQL 原子表达式（`favorite_count + 1`）避免读-改-写竞态
  - PG 额外用 SELECT ... FOR UPDATE；SQLite 依赖原子 UPDATE + WAL 写串行
  - 重复收藏：幂等（不报错，返回当前状态）
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import select, update
from sqlalchemy.orm import Session

from app.core import errors
from app.core.config import settings
from app.models import Capsule, Favorite, User
from app.schemas.capsule import FavoriteResult


def _load_capsule(db: Session, capsule_id: uuid.UUID, *, lock: bool) -> Capsule | None:
    stmt = select(Capsule).where(Capsule.id == capsule_id)
    if lock and settings.db_driver == "postgres":
        stmt = stmt.with_for_update()
    return db.execute(stmt).scalar_one_or_none()


def add_favorite(db: Session, *, user: User, capsule_id: uuid.UUID) -> FavoriteResult:
    capsule = _load_capsule(db, capsule_id, lock=True)
    if not capsule or not capsule.in_plaza:
        raise errors.not_found("胶囊不存在")
    if capsule.owner_id == user.id:
        raise errors.bad_request("不能收藏自己创建的胶囊")

    existing = db.execute(
        select(Favorite).where(Favorite.user_id == user.id, Favorite.capsule_id == capsule.id)
    ).scalar_one_or_none()
    if existing:
        return FavoriteResult(
            capsuleId=capsule.id,
            favoriteCount=capsule.favorite_count,
            favoritedAt=existing.created_at,
        )

    now = datetime.now(timezone.utc)
    db.add(Favorite(user_id=user.id, capsule_id=capsule.id, created_at=now))
    db.execute(
        update(Capsule)
        .where(Capsule.id == capsule.id)
        .values(favorite_count=Capsule.favorite_count + 1)
    )
    db.commit()
    db.refresh(capsule)
    return FavoriteResult(
        capsuleId=capsule.id, favoriteCount=capsule.favorite_count, favoritedAt=now
    )


def remove_favorite(db: Session, *, user: User, capsule_id: uuid.UUID) -> None:
    capsule = _load_capsule(db, capsule_id, lock=True)
    if not capsule:
        return
    existing = db.execute(
        select(Favorite).where(Favorite.user_id == user.id, Favorite.capsule_id == capsule.id)
    ).scalar_one_or_none()
    if not existing:
        return
    db.delete(existing)
    db.execute(
        update(Capsule)
        .where(Capsule.id == capsule.id, Capsule.favorite_count > 0)
        .values(favorite_count=Capsule.favorite_count - 1)
    )
    db.commit()

File: backends/spring-boot/src/main/java/com/hellotimepro/springboot/service/FavoriteService.java
package com.hellotimepro.springboot.service;

import com.hellotimepro.springboot.config.AppProperties;
import com.hellotimepro.springboot.domain.CapsuleEntity;
import com.hellotimepro.springboot.domain.FavoriteEntity;
import com.hellotimepro.springboot.domain.FavoriteId;
import com.hellotimepro.springboot.domain.UserEntity;
import com.hellotimepro.springboot.dto.Dtos.FavoriteResult;
import com.hellotimepro.springboot.repository.CapsuleRepository;
import com.hellotimepro.springboot.repository.FavoriteRepository;
import com.hellotimepro.springboot.web.ApiException;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class FavoriteService {
  private final AppProperties props;
  private final CapsuleRepository capsules;
  private final FavoriteRepository favorites;

  public FavoriteService(AppProperties props, CapsuleRepository capsules, FavoriteRepository favorites) {
    this.props = props;
    this.capsules = capsules;
    this.favorites = favorites;
  }

  @Transactional
  public FavoriteResult addFavorite(UserEntity user, String capsuleId) {
    // favorite_count 是冗余计数器，必须和 favorites 行变更处在同一事务。
    // Postgres 路径会拿胶囊行写锁；SQLite 依赖单写事务和原子 UPDATE。
    CapsuleEntity capsule = lockCapsule(capsuleId);
    if (!capsule.isInPlaza()) throw ApiException.notFound("胶囊不存在");
    if (capsule.getOwnerId().equals(user.getId())) throw ApiException.badRequest("不能收藏自己创建的胶囊");

    FavoriteId id = new FavoriteId(user.getId(), capsule.getId());
    FavoriteEntity existing = favorites.findById(id).orElse(null);
    if (existing != null) {
      return new FavoriteResult(capsule.getId().toString(), capsule.getFavoriteCount(), existing.getCreatedAt());
    }

    OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
    FavoriteEntity favorite = new FavoriteEntity();
    favorite.setId(id);
    favorite.setCreatedAt(now);
    favorites.save(favorite);
    capsules.incrementFavoriteCount(capsule.getId());
    int newCount = capsules.findById(capsule.getId())
        .map(CapsuleEntity::getFavoriteCount)
        .orElse(capsule.getFavoriteCount() + 1);
    return new FavoriteResult(capsule.getId().toString(), newCount, now);
  }

  @Transactional
  public void removeFavorite(UserEntity user, String capsuleId) {
    // 取消收藏同样在事务里删除 favorites 行并递减计数，保持 spec 的 I2 不变式。
    UUID uuid = UUID.fromString(capsuleId);
    CapsuleEntity capsule = capsules.findById(uuid).orElse(null);
    if (capsule == null) return;
    FavoriteId id = new FavoriteId(user.getId(), capsule.getId());
    if (favorites.existsById(id)) {
      favorites.deleteById(id);
      capsules.decrementFavoriteCount(capsule.getId());
    }
  }

  private CapsuleEntity lockCapsule(String id) {
    UUID uuid = UUID.fromString(id);
    if ("postgres".equals(props.getDbDriver())) {
      // JPA 的 PESSIMISTIC_WRITE 对应 SELECT ... FOR UPDATE，避免并发收藏时计数漂移。
      return capsules.findByIdForUpdate(uuid).orElseThrow(() -> ApiException.notFound("胶囊不存在"));
    }
    return capsules.findById(uuid).orElseThrow(() -> ApiException.notFound("胶囊不存在"));
  }
}
